File: src/platform/posix/random.c
/*
 * Using Chacha as streamcipher, keyed per thread and getting
 * seed from OS specific methods and a fallback to /dev/urandom
 * similar to arc4random and friends, of course, but given that
 */

#include "chacha.c"
#include <stdio.h>
#include <errno.h>
#include <sys/param.h>

void arcan_fatal(const char* msg, ...);

static _Thread_local struct chacha_ctx streamcipher;

/* just lifted from the getrandom manpage, on others
 * (osx, fbsd, openbsd) we assume that we have this one */
#if defined(__LINUX) || defined(__DragonFly__)

/* normally defined in unistd.h, but that one also pulls in
 * getentropy of its own, though not always available..., */
long syscall(long number, ...);
#include <sys/syscall.h>
static int seedrnd(void* buf, size_t buflen)
{
	int ret;
	if (buflen > 256)
		goto failure;

#ifndef SYS_getrandom
#else
	ret = syscall(SYS_getrandom, buf, buflen, 0);
	if (ret < 0)
		return ret;
	if (ret == buflen)
		return 0;
#endif

failure:
	errno = EIO;
	return -1;
}
#elif defined(__FreeBSD__) && __FreeBSD__ < 12
#include <sys/sysctl.h>
extern int __sysctl(int*, u_int, void*, size_t*, void*, size_t);
static int seedrnd(void* buf, size_t buflen)
{
	int mib[2] = {CTL_KERN, KERN_ARND};
	size_t size = buflen, len;
	uint8_t* db = (uint8_t*) buf;

	while(size){
		len = size;
		if (__sysctl(mib, 2, db, &len, NULL, 0) == -1){
			goto out;
		}
		db += len;
		size -= len;
	}

out:
	if (size != 0){
		errno = EIO;
		return -1;
	}
	return 0;
}

#elif defined(__FreeBSD__) || defined(__OpenBSD__)
#include <unistd.h>
#define seedrnd getentropy
#else
#include <sys/random.h>
#define seedrnd getentropy
#endif

static void seed_csprng()
{
	uint8_t seed[16];
	uint8_t nonce[8] = {9};

/* fallback, /dev/urandom - hopefully it will never be used as it would need to
 * be taken into account when it comes to chroot/jail/privsep */
	if (seedrnd(seed, 16) != 0){
		FILE* fpek = fopen("/dev/urandom", "r");
		if (!fpek || 1 != fread(seed, 16, 1, fpek))
			arcan_fatal("couldn't seed CSPRNG, system not in a safe state\n");
		fclose(fpek);
	}

	chacha_setup(&streamcipher, seed, 16, nonce, 0, 8);
}

void arcan_random(uint8_t* dst, size_t ntc)
{
	if (!streamcipher.ready)
		seed_csprng();

/* slow-copy until we're aligned */
	uint32_t ibuf[16];
	size_t ofs = 0;
	size_t malign = (uintptr_t) dst % sizeof(uint32_t*);
	if (malign && ntc >= 64){
		chacha_block(&streamcipher, ibuf);
		memcpy(dst, ibuf, malign);
		ofs += malign;
		ntc -= malign;
	}

/* as we're only using the keystream as a csprng, any leftover bytes
 * in each block can be ignored */
	while (ntc >= 64){
		chacha_block(&streamcipher, (uint32_t*)&dst[ofs]);
		ofs += 64;
		ntc -= 64;
	}

	if (!ntc)
		return;

/* and slow-copy the remaining bytes */
	chacha_block(&streamcipher, ibuf);
	memcpy(&dst[ofs], ibuf, ntc);
}

File: src/platform/posix/random.zig
// Pure Zig port of posix/random.c
// Thread-local ChaCha8 CSPRNG seeded from OS entropy (getrandom).
// Replaces the C chacha.c implementation with std.crypto.stream.chacha.

const std = @import("std");
const builtin = @import("builtin");
const is_freestanding = (builtin.os.tag == .freestanding);

extern fn arcan_fatal(msg: [*c]const u8, ...) callconv(.c) void;

const ChaCha8IETF = std.crypto.stream.chacha.ChaCha8IETF;

/// Thread-local CSPRNG state: a ChaCha8 cipher seeded lazily from OS entropy.
const CsprngState = struct {
    ready: bool = false,
    /// Buffered keystream for partial-block requests.
    buf: [64]u8 = undefined,
    buf_pos: usize = 64, // start exhausted so first call generates a block
    /// ChaCha8 counter-mode state: key + nonce + block counter.
    key: [32]u8 = undefined,
    nonce: [12]u8 = undefined,
    counter: u32 = 0,

    fn seed(self: *CsprngState) void {
        var seed_buf: [32]u8 = undefined;
        if (is_freestanding) {
            // On freestanding, zero-seed (no OS entropy available)
            @memset(&seed_buf, 0);
        } else {
            seed_from_os(&seed_buf);
        }

        self.key = seed_buf;
        self.nonce = .{ 9, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
        self.counter = 0;
        self.buf_pos = 64; // mark buffer as exhausted
        self.ready = true;
    }

    /// Generate one 64-byte keystream block into `out`.
    fn generateBlock(self: *CsprngState, out: *[64]u8) void {
        // ChaCha8IETF.xor produces keystream when XOR'd with zeros
        const zeros: [64]u8 = .{0} ** 64;
        ChaCha8IETF.xor(out, &zeros, self.counter, self.key, self.nonce);
        self.counter +%= 1;
    }

    /// Fill `dst` with random bytes from the keystream.
    fn fill(self: *CsprngState, dst: []u8) void {
        var remaining = dst;

        // First, drain any leftover bytes from the internal buffer
        if (self.buf_pos < 64) {
            const avail = 64 - self.buf_pos;
            const take = @min(avail, remaining.len);
            @memcpy(remaining[0..take], self.buf[self.buf_pos..][0..take]);
            self.buf_pos += take;
            remaining = remaining[take..];
        }

        // Generate full 64-byte blocks directly into the output
        while (remaining.len >= 64) {
            self.generateBlock(remaining[0..64]);
            remaining = remaining[64..];
        }

        // Handle the tail: generate a block, copy what we need, buffer the rest
        if (remaining.len > 0) {
            self.generateBlock(&self.buf);
            @memcpy(remaining, self.buf[0..remaining.len]);
            self.buf_pos = remaining.len;
        }
    }
};

fn seed_from_os(seed_buf: *[32]u8) void {
    // Try OS CSPRNG (getrandom syscall on Linux)
    std.posix.getrandom(seed_buf) catch {
        // Fallback: /dev/urandom
        const f = std.fs.openFileAbsolute("/dev/urandom", .{}) catch {
            arcan_fatal("couldn't seed CSPRNG, system not in a safe state\n");
            unreachable;
        };
        defer f.close();
        const nread = f.read(seed_buf) catch {
            arcan_fatal("couldn't seed CSPRNG, system not in a safe state\n");
            unreachable;
        };
        if (nread != seed_buf.len) {
            arcan_fatal("couldn't seed CSPRNG, system not in a safe state\n");
            unreachable;
        }
    };
}

threadlocal var state: CsprngState = .{};

export fn arcan_random(dst: [*]u8, ntc: usize) callconv(.c) void {
    if (!state.ready) state.seed();
    state.fill(dst[0..ntc]);
}
